### PR TITLE
Manage excluded_products in ranges dashboard

### PR DESCRIPTION
**Pull request by gwaidacher**
Added tabs to range - products in dashboard, first tab contains now existing form for managing included products, in the second tab the excluded products can be managed, with a separate textfield for upc input and a second fileupload form.

see https://github.com/django-oscar/django-oscar/issues/4055

### CODE REVIEW

**Comment by gwaidacher** [comment]
Added missing tests. where does the "1 file has unexpected coverage changes not visible in diff " error come from?

**Review by joeyjurjens — COMMENTED** [review]
Hi @gwaidacher , thank you very much for your PR. It's quite a big one and we really appreciate it that you're taking the time for it!

I do have some review comments though, if you have any questions or comments to mine, just let me know! :)

File: src/oscar/apps/dashboard/ranges/views.py [hunk 1]
```diff
@@ -130,41 +148,86 @@ def remove_selected_products(self, request, products):
...
+        if not form.is_valid():
+            ctx = self.get_context_data(form_excluded=form,
+                                        object_list=self.object_list)
+            ctx['active_tab'] = 'excluded'
+            return self.render_to_response(ctx)
+
+        self.handle_query_products(request, range, form)
+        self.handle_file_products(request, range, form, included=False)
+        return redirect(reverse('dashboard:range-products-excluded',
+                                kwargs={'pk': range.pk, 'excluded': 1}))
```
[the remark this answers is not in the document]
Comment (gwaidacher) [reply src/oscar/apps/dashboard/ranges/views.py:185]: It's used to open the 2nd Tab with the excluded products when the page is loaded.

File: src/oscar/views/generic.py [hunk 2]
```diff
@@ -57,8 +57,8 @@ def get_checkbox_object_name(self):
     def get_error_url(self, request):
         return safe_referrer(request, '.')
 
-    def get_success_url(self, request):
-        return safe_referrer(request, '.')
+    # def get_success_url(self, request):
```
[the remark this answers is not in the document]
Comment (gwaidacher) [reply src/oscar/views/generic.py:60]: It is not used any more which lead to the "1 file has unexpected coverage changes not visible in diff " Error in codecov report.

File: src/oscar/apps/dashboard/ranges/views.py [hunk 3]
```diff
@@ -96,14 +95,24 @@ class RangeProductListView(BulkEditMixin, ListView):
...
     template_name = 'oscar/dashboard/ranges/range_product_list.html'
     context_object_name = 'products'
-    actions = ('remove_selected_products', 'add_products')
+    actions = ('add_products', 'add_excluded_products',
+               'remove_selected_products', 'remove_excluded_products')
     form_class = RangeProductForm
     paginate_by = settings.OSCAR_DASHBOARD_ITEMS_PER_PAGE
 
+    def get(self, request, **kwargs):
+        self.excluded = 'excluded' in kwargs
```
[the remark this answers is not in the document]
Comment (gwaidacher) [reply src/oscar/apps/dashboard/ranges/views.py:104]: its used to switch to the 2nd Tab with the excluded_products (see ctx['active_tab'] in get_context_data() and on form.in_valid() in add_excluded_products() )

File: src/oscar/apps/dashboard/ranges/views.py [hunk 4]
```diff
@@ -130,41 +148,86 @@ def remove_selected_products(self, request, products):
...
-        self.handle_file_products(request, range, form)
-        return HttpResponseRedirect(self.get_success_url(request))
+        self.handle_file_products(request, range, form, included=True)
+        return redirect(reverse('dashboard:range-products',
+                                kwargs={'pk': range.pk}))
+
+    def add_excluded_products(self, request):
+        range = self.get_range()
+        form = self.form_class(range, request.POST, request.FILES)
+        form.included = False
```
[the remark this answers is not in the document]
Comment (gwaidacher) [reply src/oscar/apps/dashboard/ranges/views.py:175]: I have tried it before with the hidden boolean input and passing it via argument through the __init__ method but I received form errors with it which I could not resolve. My impatience made me solve it this way, though I know it's not the django way...

File: src/oscar/apps/offer/abstract_models.py [hunk 5]
```diff
@@ -1042,6 +1042,8 @@ class AbstractRangeProductFileUpload(models.Model):
         on_delete=models.CASCADE,
         verbose_name=_("Uploaded By"))
     date_uploaded = models.DateTimeField(_("Date Uploaded"), auto_now_add=True, db_index=True)
+    # for managing excluded_products
+    included = models.BooleanField(default=True)
```
[the remark this answers is not in the document]
Comment (gwaidacher) [reply src/oscar/apps/offer/abstract_models.py]: Hi! I restructured this PR fitting your suggestions, please have a look!

**Comment by gwaidacher** [comment]
* changed field "included to "upload_type" in RangeProductFileUpload
* changed the "excluded" kwarg in redirect url to "active_tab"
* passed the "excluded" argument in RangeProductForm via __init__() method

File: src/oscar/apps/offer/abstract_models.py [hunk 6]
```diff
@@ -1090,9 +1092,13 @@ def was_processing_successful(self):
     def process(self, file_obj):
         """
         Process the file upload and add products to the range
+        or add products to range.excluded_products
         """
         all_ids = set(self.extract_ids(file_obj))
-        products = self.range.all_products()
+        if self.included:
```
[the remark this answers is not in the document]
Comment (gwaidacher) [reply src/oscar/apps/offer/abstract_models.py]: Done.

[the remark this answers is not in the document]
Comment (gwaidacher) [reply src/oscar/apps/offer/abstract_models.py]: done.

**Comment by gwaidacher** [comment]
> Hi @gwaidacher , thank you very much for your PR. It's quite a big one and we really appreciate it that you're taking the time for it!
> 
> I do have some review comments though, if you have any questions or comments to mine, just let me know! :)

PR updated, please have a look

File: src/oscar/apps/dashboard/ranges/forms.py [hunk 7]
```diff
@@ -31,8 +31,9 @@ class RangeProductForm(forms.Form):
         label=_("File of SKUs or UPCs"), required=False, max_length=255,
         help_text=_('Either comma-separated, or one identifier per line'))
 
-    def __init__(self, range, *args, **kwargs):
+    def __init__(self, range, excluded, *args, **kwargs):
```
Comment (viggo-devries) [inline src/oscar/apps/dashboard/ranges/forms.py:34]: please make this a kwarts, excluded=True/False, otherwise it's really hard to know in the view what this does.

**Review by viggo-devries — CHANGES_REQUESTED** [review]

File: src/oscar/apps/dashboard/ranges/forms.py (hunk 7 above)
Comment (gwaidacher) [reply src/oscar/apps/dashboard/ranges/forms.py:34]: done.

Comment (viggo-devries) [reply src/oscar/apps/dashboard/ranges/forms.py:34]: Yes, and then you can remove it from the __init__ args.

Comment (gwaidacher) [reply src/oscar/apps/dashboard/ranges/forms.py]: done.

File: src/oscar/apps/dashboard/ranges/apps.py [hunk 8]
```diff
@@ -27,6 +27,8 @@ def get_urls(self):
             path('<int:pk>/', self.update_view.as_view(), name='range-update'),
             path('<int:pk>/delete/', self.delete_view.as_view(), name='range-delete'),
             path('<int:pk>/products/', self.products_view.as_view(), name='range-products'),
+            path('<int:pk>/products/<int:active_tab>/', self.products_view.as_view(),
```
Comment (joeyjurjens) [inline src/oscar/apps/dashboard/ranges/apps.py:30]: I still don't think this is necessary. If you want to remember what the active tab was before submiting the form, I suggest using a new HiddenFormField which you initially set on the forms.

File: src/oscar/apps/dashboard/ranges/forms.py [hunk 9]
```diff
@@ -33,6 +33,7 @@ class RangeProductForm(forms.Form):
 
     def __init__(self, range, *args, **kwargs):
         self.range = range
+        self.excluded = kwargs.pop('excluded', False)
```
Comment (joeyjurjens) [inline src/oscar/apps/dashboard/ranges/forms.py:36]: If you add a new hidden input field, which describes what kind of form it is, you can just remove this.
Inside the clean method you could just access the hidden form field and do logic based on that.

```python
upload_type = forms.CharField(widget=forms.HiddenInput())
```

File: src/oscar/apps/dashboard/ranges/views.py [hunk 10]
```diff
@@ -96,14 +95,24 @@ class RangeProductListView(BulkEditMixin, ListView):
     model = Product
     template_name = 'oscar/dashboard/ranges/range_product_list.html'
     context_object_name = 'products'
-    actions = ('remove_selected_products', 'add_products')
+    actions = ('add_products', 'add_excluded_products',
+               'remove_selected_products', 'remove_excluded_products')
     form_class = RangeProductForm
     paginate_by = settings.OSCAR_DASHBOARD_ITEMS_PER_PAGE
 
+    def get(self, request, **kwargs):
```
Comment (joeyjurjens) [inline src/oscar/apps/dashboard/ranges/views.py:103]: With the hidden form field, you can also remove this

File: src/oscar/apps/dashboard/ranges/views.py [hunk 11]
```diff
@@ -96,14 +95,24 @@ class RangeProductListView(BulkEditMixin, ListView):
...
+        self.active_tab = kwargs.get('active_tab', 0)
+        return super().get(request, **kwargs)
+
     def post(self, request, *args, **kwargs):
+        self.active_tab = kwargs.get('active_tab', 0)
```
Comment (joeyjurjens) [inline src/oscar/apps/dashboard/ranges/views.py:108]: With the hidden form field, you can also remove this

File: src/oscar/apps/dashboard/ranges/views.py [hunk 12]
```diff
@@ -121,6 +130,13 @@ def get_context_data(self, **kwargs):
         ctx['range'] = range
         if 'form' not in ctx:
             ctx['form'] = self.form_class(range)
```
Comment (joeyjurjens) [inline src/oscar/apps/dashboard/ranges/views.py:132]: Here, you would set the intiial value of the upload type. Say you named the hidden field: upload_type, you can do this:

```python
ctx['form'] = self.form_class(range, initial={"upload_type": INCLUDED_PRODUCTS_TYPE})
```

File: src/oscar/apps/dashboard/ranges/views.py [hunk 13]
```diff
@@ -121,6 +130,13 @@ def get_context_data(self, **kwargs):
...
+        if 'form_excluded' not in ctx:
+            ctx['form_excluded'] = self.form_class(range, excluded=True)
```
Comment (joeyjurjens) [inline src/oscar/apps/dashboard/ranges/views.py:134]: And here this would be:

```python
ctx['form_excluded'] = self.form_class(range, initial={"upload_type": EXCLUDED_PRODUCTS_TYPE})
```

File: src/oscar/apps/dashboard/ranges/views.py [hunk 14]
```diff
@@ -143,28 +162,68 @@ def add_products(self, request):
...
         self.handle_query_products(request, range, form)
-        self.handle_file_products(request, range, form)
-        return HttpResponseRedirect(self.get_success_url(request))
+        self.handle_file_products(request, range, form, included=True)
+        return redirect(reverse('dashboard:range-products',
+                                kwargs={'pk': range.pk}))
+
+    def add_excluded_products(self, request):
+        range = self.get_range()
+        form = self.form_class(
```
Comment (joeyjurjens) [inline src/oscar/apps/dashboard/ranges/views.py:171]: Here you can remove the `excluded=True` as you have access to the upload type:
form.cleaned_data["upload_type"] and thus you also know what the active tab will be.

File: src/oscar/apps/dashboard/ranges/views.py [hunk 15]
```diff
@@ -174,11 +233,12 @@ def handle_query_products(self, request, range, form):
                 ", ".join(missing_skus))
         self.check_imported_products_sku_duplicates(request, products)
 
-    def handle_file_products(self, request, range, form):
+    def handle_file_products(self, request, range, form, included=True):
```
Comment (joeyjurjens) [inline src/oscar/apps/dashboard/ranges/views.py:236]: Again, you could remove included=True here because you can access the type from the `form.cleaned_data["upload_type"]`

File: src/oscar/apps/dashboard/ranges/views.py [hunk 16]
```diff
@@ -190,12 +250,17 @@ def handle_file_products(self, request, range, form):
             messages.success(request, msg, extra_tags='safe noicon block')
         self.check_imported_products_sku_duplicates(request, products)
 
-    def create_upload_object(self, request, range, f):
+    def create_upload_object(self, request, range, f, included=True):
```
Comment (joeyjurjens) [inline src/oscar/apps/dashboard/ranges/views.py:253]: Same as above, you can remove included here, because you have the upload type in `f.cleaned_data["upload_type"]`

File: src/oscar/views/generic.py (hunk 2 above)
Comment (joeyjurjens) [inline src/oscar/views/generic.py:60]: I still don't see how this method has anything to do with this PR? Can't you just keep it like it was?

File: tests/functional/dashboard/test_range.py [hunk 17]
```diff
@@ -20,7 +20,8 @@ def tearDown(self):
         Range.objects.all().delete()
 
     def submit_form(self, data):
-        return forms.RangeProductForm(self.range, data)
+        form = forms.RangeProductForm(self.range, data)
```
Comment (joeyjurjens) [inline tests/functional/dashboard/test_range.py:23]: Is this change really necessary?

**Review by joeyjurjens — CHANGES_REQUESTED** [review]
Hi @gwaidacher , I still have some feedback to keep it a bit more clean. Thanks again for your time and effort.

File: src/oscar/views/generic.py (hunk 2 above)
Comment (gwaidacher) [reply src/oscar/views/generic.py:60]: There was a codecov error concerning this function, so I commented it out to have all checks passed. It's not needed any more by ranges.views.py

File: src/oscar/apps/dashboard/ranges/forms.py [hunk 18]
```diff
@@ -30,9 +31,15 @@ class RangeProductForm(forms.Form):
     file_upload = forms.FileField(
         label=_("File of SKUs or UPCs"), required=False, max_length=255,
         help_text=_('Either comma-separated, or one identifier per line'))
+    upload_type = forms.CharField(widget=forms.HiddenInput(), required=False)
 
     def __init__(self, range, *args, **kwargs):
         self.range = range
+        # make upload_type accessible in clean_query
+        if kwargs.get("initial"):
```
Comment (gwaidacher) [inline src/oscar/apps/dashboard/ranges/forms.py:39]: I cannot access cleaned_data['upload_type'] in clean_query() - the solution to avoid this would be to merge clean_query() into clean()

File: src/oscar/apps/dashboard/ranges/views.py [hunk 19]
```diff
@@ -144,27 +166,73 @@ def add_products(self, request):
...
+        if not form.is_valid():
+            ctx = self.get_context_data(form_excluded=form,
+                                        object_list=self.object_list)
+            return self.render_to_response(ctx)
+
+        self.handle_query_products(request, range, form)
+        self.handle_file_products(request, range, form)
+        return redirect(
+            reverse('dashboard:range-products', kwargs={'pk': range.pk})
+            + '?upload_type=excluded'
```
Comment (joeyjurjens) [inline src/oscar/apps/dashboard/ranges/views.py:188]: I don't think this is really necessary. Once the upload was successful, we can just redirect them as is, maybe with a success message stating it was successful. The GET parameter is not really needed in my opinion.

**Review by joeyjurjens — CHANGES_REQUESTED** [review]
Few small things left, the rest looks great!

File: src/oscar/apps/dashboard/ranges/forms.py [hunk 20]
```diff
@@ -40,43 +42,47 @@ def clean(self):
...
-            'stockrecords__partner_sku', flat=True))
-        existing_upcs = set(products.values_list('upc', flat=True))
-        existing_ids = existing_skus.union(existing_upcs)
-        new_ids = ids - existing_ids
-
-        if len(new_ids) == 0:
-            raise forms.ValidationError(
-                _("The products with SKUs or UPCs matching %s are already in"
-                  " this range") % (', '.join(ids)))
+        raw = clean_data['query']
```
[the remark this answers is not in the document]
Comment (gwaidacher) [reply src/oscar/apps/dashboard/ranges/forms.py:45]: It was not possible to add the changes in clean_query, because I could not access `clean_data['upload_type']` there. Otherwise I would have to call super.clean() twice. Any ideas?

File: src/oscar/apps/dashboard/ranges/views.py (hunk 19 above)
Comment (gwaidacher) [reply src/oscar/apps/dashboard/ranges/views.py:188]: It's used to redirected to the page having the second tab with excluded products active.

File: src/oscar/apps/dashboard/ranges/forms.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (gwaidacher) [reply src/oscar/apps/dashboard/ranges/forms.py:82]: ah, misunderstood... fixed.

File: src/oscar/templates/oscar/dashboard/ranges/range_product_list.html [hunk 21]
```diff
@@ -30,114 +30,248 @@
...
+          {% if range.excluded_products.count %}
+            <form method="post">
+              {% csrf_token %}
+              <table class="table table-striped table-bordered table-hover">
+               <caption>
+                  <h3 class="float-left">{% trans "Products excluded from range" %}</h3>
+                  <div class="float-right">
+                    <input type="hidden" name="action" value="remove_excluded_products" />
+                    <button type="submit" class="btn btn-secondary" data-loading-text="{% trans 'Removing...' %}">
+                      {% trans "Remove excluded products" %}
```
Comment (viggo-devries) [inline src/oscar/templates/oscar/dashboard/ranges/range_product_list.html:224]: Please change this to:

Remove selected excluded products

File: src/oscar/apps/dashboard/ranges/views.py [hunk 22]
```diff
@@ -144,27 +166,73 @@ def add_products(self, request):
...
         products = form.get_products()
         if not products:
             return
-
         for product in products:
-            range.add_product(product)
-
+            if form.cleaned_data["upload_type"] == \
+                    RangeProductFileUpload.EXCLUDED_PRODUCTS_TYPE:
+                action = _('excluded from this range')
```
[the remark this answers is not in the document]
Comment (viggo-devries) [reply src/oscar/apps/dashboard/ranges/views.py:217]: please add a test for this

**Review by viggo-devries — CHANGES_REQUESTED** [review]

**Comment by viggo-devries** [comment]
In general this is missing some tests, great that the existing tests still pass but please add some specific tests for excluding the products.

File: src/oscar/templates/oscar/dashboard/ranges/range_product_list.html (hunk 21 above)
Comment (gwaidacher) [reply src/oscar/templates/oscar/dashboard/ranges/range_product_list.html]: fixed

**Comment by gwaidacher** [comment]
> In general this is missing some tests, great that the existing tests still pass but please add some specific tests for excluding the products.

Hi! There are already 2 tests in test_range.RangeProductViewTest: test_upload_excluded_file_with_skus for the upload-file form and test_dupe_excluded_skus_warning - if this is insufficient, please specify which function should be tested.

**Comment by viggo-devries** [comment]
> > In general this is missing some tests, great that the existing tests still pass but please add some specific tests for excluding the products.
> 
> Hi! There are already 2 tests in test_range.RangeProductViewTest: test_upload_excluded_file_with_skus for the upload-file form and test_dupe_excluded_skus_warning - if this is insufficient, please specify which function should be tested.

Yes I saw but the fact that your tests missed this entire functionality: https://github.com/django-oscar/django-oscar/pull/4073/commits/f97053a741a3a65d35b8d45d563c46480c500325 is not good.

I would like to see tests for every action that can be performed and also the result of that action.
You should be able to come up with way more than the 2 actions you have tested now. like:
test_add_single_product_to_excluded_range
test_add_multipe_products_to_excluded_range
remove_single_product_from_excluded_range
remove_multiple_products_from_excluded_range
etc.etc.

**Comment by joeyjurjens** [comment]
@gwaidacher Hey man. I have written the missing tests, so you don't have to do that anymore.

**Comment by gwaidacher** [comment]
> @gwaidacher Hey man. I have written the missing tests, so you don't have to do that anymore.

Thank you very much :D

**Review by joeyjurjens — APPROVED** [review]
